File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/KFactory.java
package org.kevoree.modeling.api;

import org.kevoree.modeling.api.json.JSONModelLoader;
import org.kevoree.modeling.api.json.JSONModelSerializer;
import org.kevoree.modeling.api.meta.MetaClass;
import org.kevoree.modeling.api.xmi.XMIModelLoader;
import org.kevoree.modeling.api.xmi.XMIModelSerializer;

import java.util.List;

/**
 * Created by thomas on 10/2/14.
 */
public interface KFactory {

    public KObject create(String metaClassName);

    public void root(KObject elem, Callback<Boolean> callback);

    public ModelSerializer createJSONSerializer();

    public ModelLoader createJSONLoader();

    public ModelSerializer createXMISerializer();

    public ModelLoader createXMILoader();

    public ModelCompare createModelCompare();

    public ModelCloner createModelCloner();

    public ModelSlicer createModelSlicer();

    public void select(String query, Callback<List<KObject>> callback);

    public void lookup(String path, Callback<KObject> callback);

    public void stream(String query, Callback<KObject> callback);

    public MetaClass[] metaClasses();

}

File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/KFactory.kt
package org.kevoree.modeling.api

import org.kevoree.modeling.api.json.JSONModelSerializer
import org.kevoree.modeling.api.json.JSONModelLoader
import org.kevoree.modeling.api.xmi.XMIModelLoader
import org.kevoree.modeling.api.xmi.XMIModelSerializer
import org.kevoree.modeling.api.compare.DefaultModelCompare

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 05/08/13
 * Time: 21:00
 */

public trait KFactory {

    fun create(metaClassName: String, callback: Callback<org.kevoree.modeling.api.KObject<*>?>)

    fun root(elem: KObject<*>, callback: Callback<Boolean>)

    fun createJSONSerializer(): JSONModelSerializer

    fun createJSONLoader(): JSONModelLoader

    fun createXMISerializer(): XMIModelSerializer

    fun createXMILoader(): XMIModelLoader

    fun createModelCompare(): ModelCompare

    fun createModelCloner(): ModelCloner

    fun createModelSlicer(): ModelSlicer

    fun select(query: String, callback: Callback<List<KObject<*>>>)

    fun lookup(path: String, callback: Callback<KObject<*>>)

    fun stream(query: String, callback: Callback<KObject<*>>)

}
